### PR TITLE
Feature/issue 20 leaderboard

### PR DESCRIPTION
**Pull request by jaderabbit**
Fixed the leaderboard to match SoW description: Only show Top 10

### CODE REVIEW

File: oneplus/templates/prog/leader.html [hunk 1]
```diff
@@ -27,7 +31,7 @@
 
         {% for participant in learners %}
             <div class="item grey-back {% if participant.me %}red-front{% endif %}">
-                <div class="large-font pull-left">#1</div>
+                <div class="large-font pull-left">{{ forloop.counter }}</div>
```
Comment (smn) [inline oneplus/templates/prog/leader.html]: Not sure if omitting the `#` was intentional (or originally accidental)?

File: oneplus/views.py [hunk 2]
```diff
@@ -818,20 +818,39 @@ def leader(request, state, user):
     _participant.me = True
     request.session["state"]["leader_region"] = "Countrywide"
 
+    def leader_position(location):
+        if location == "Countrywide":
```
Comment (smn) [inline oneplus/views.py]: Can we turn this into a `CONSTANT` somewhere?

File: oneplus/views.py [hunk 3]
```diff
@@ -818,20 +818,39 @@ def leader(request, state, user):
...
+            return Participant.objects.filter(classs=_participant.classs, points__gt=_participant.points).count() + 1
```
Comment (smn) [inline oneplus/views.py]: `classs` is asking for trouble, either make it obvious and go for `klass` or rather just name it `ParticipantClass`.

File: oneplus/views.py [hunk 4]
```diff
@@ -818,20 +818,39 @@ def leader(request, state, user):
     _participant.me = True
```
Comment (smn) [inline oneplus/views.py]: would it make sense to have `request.user` grow a `get_participant()` helper function? I'm guessing you're going to be needing it in a bunch of places and will probably eliminate the need for this fairly brittle `.me = True` flag.

File: oneplus/views.py [hunk 5]
```diff
@@ -818,20 +818,39 @@ def leader(request, state, user):
...
 
+    def leader_position(location):
+        if location == "Countrywide":
+            return Participant.objects.filter(classs=_participant.classs, points__gt=_participant.points).count() + 1
+
+        if _participant.learner.area == location:
+            return Participant.objects.filter(
+                classs=_participant.classs,
+                points__gt=_participant.points,
+                learner__area=location).count() + 1
```
Comment (smn) [inline oneplus/views.py]: Postgres is terrible at counting, it's _seriously_ slow. Iirc this is because of its MVCC nature and as a result often needs to do a table scan to eliminate tombstones in order to do a full count. Once this site hits 100K+ users performance is going to tank. 

If we're going to need this often can we find a way of doing this without requiring Postgres to do a `count()`? Can we get away with an incrementing participants integer on a location somewhere?

File: oneplus/views.py [hunk 6]
```diff
@@ -818,20 +818,39 @@ def leader(request, state, user):
...
+        else:
+            return -1
+
+    def get_leaderboard(location):
+        if location == "Countrywide":
+            return list(Participant.objects.filter(classs=_participant.classs).order_by("-points")[:10])
+        else:
+            return list(Participant.objects.filter(
+                        classs=_participant.classs,
+                        learner__area=location).order_by("-points")[:10])
```
Comment (smn) [inline oneplus/views.py]: Rather not `list()` these. Doing so always has Postgres execute the query, returning it gives you the Query object on top of which you can continue chaining stuff.

File: oneplus/views.py (hunk 2 above)
Comment (jaderabbit) [inline oneplus/views.py]: Any particular preference as to where?

File: oneplus/views.py (hunk 5 above)
Comment (jaderabbit) [inline oneplus/views.py]: Ugh. How will I live with myself? >.<

File: oneplus/views.py (hunk 3 above)
Comment (jaderabbit) [inline oneplus/views.py]: Going to put this is another PR since it requires changes to mobileu

File: oneplus/views.py (hunk 4 above)
Comment (jaderabbit) [inline oneplus/views.py]: That was just so the templates originally knew which user to highlight in the leaderboard list. But I do that later in the code in a different way now so that line can actually be removed.

File: oneplus/views.py (hunk 5 above)
Comment (jaderabbit) [inline oneplus/views.py]: Also, will probably need to put this in another PR. It will probably need to be sorted mobileu side

**Comment by jaderabbit** [comment]
@smn Ready for review

File: oneplus/views.py [hunk 7]
```diff
@@ -226,43 +266,76 @@ def post():
...
             _learnerstate.active_result = _option.correct
             _learnerstate.save()
 
             #Check for awards
             if _option.correct:
-                _participant.award_scenario("CORRECT", _learnerstate.active_question.bank.module)
-
-                _total_corect = ParticipantQuestionAnswer.objects.filter(participant=_participant, correct=True).count()
+                _participant.award_scenario(
+                    "CORRECT",
```
Comment (smn) [inline oneplus/views.py:286]: should this perhaps be an attribute on the Participant model? `Participant.CORRECT` or something similar?

**Comment by smn** [comment]
:+1: on this landing because of all the pep8 changes, makes it a bit hard to see what's going on.
I am still concerned about the massive queries being generated, but that concern will probably still linger for a while. Let's get to the point where we can load test and see how things break.

File: oneplus/views.py (hunk 7 above)
Comment (jaderabbit) [inline oneplus/views.py:286]: I'll create an issue for mobileu.
